File: Chapter1/Codes_Figures/spurious_regression.py
"""
Spurious Regression: Nine Regressions of Independent Random Walks
==================================================================
Author:   Alessia Paccagnini
Textbook: Macroeconometrics
"""

import numpy as np
import matplotlib.pyplot as plt
import statsmodels.api as sm
from statsmodels.stats.stattools import durbin_watson
import os

np.random.seed(None)  # We'll try seeds until we get good results

# Find 9 pairs with R2 > 0.80 and DW < 0.20
T = 300
results = []
max_attempts = 100000
attempt = 0

while len(results) < 9 and attempt < max_attempts:
    attempt += 1
    # Generate two independent random walks
    x = np.cumsum(np.random.randn(T))
    y = np.cumsum(np.random.randn(T))
    
    # Regress y on x
    X = sm.add_constant(x)
    model = sm.OLS(y, X).fit()
    r2 = model.rsquared
    t_stat = model.tvalues[1]
    dw = durbin_watson(model.resid)
    
    if r2 > 0.80 and dw < 0.20:
        results.append({
            'x': x, 'y': y,
            'r2': r2, 't': t_stat, 'dw': dw,
            'fitted': model.fittedvalues
        })
        print(f"Found pair {len(results)}: R2={r2:.2f}, t={t_stat:.1f}, DW={dw:.2f}")

print(f"\nTotal attempts: {attempt}")
print(f"Pairs found: {len(results)}")

# Plot
fig, axes = plt.subplots(3, 3, figsize=(14, 12))
fig.suptitle('Nine Regressions of Independent Random Walks\n(All relationships are SPURIOUS)',
             fontsize=14, fontweight='bold', y=0.98)

for i, (ax, res) in enumerate(zip(axes.flatten(), results)):
    # Scatter plot
    ax.scatter(res['x'], res['y'], alpha=0.45, s=15, color='#9B59B6', edgecolors='#7D3C98', linewidth=0.3)
    
    # Regression line
    sort_idx = np.argsort(res['x'])
    ax.plot(res['x'][sort_idx], res['fitted'][sort_idx], color='red', linewidth=2)
    
    # Format t-stat with significance stars
    abs_t = abs(res['t'])
    if abs_t > 3.29:
        stars = '***'
    elif abs_t > 2.58:
        stars = '**'
    elif abs_t > 1.96:
        stars = '*'
    else:
        stars = ''
    
    t_sign = '-' if res['t'] < 0 else ''
    title = f"$R^2$={res['r2']:.2f}, t={t_sign}{abs_t:.1f}{stars}, DW={res['dw']:.2f}"
    ax.set_title(title, fontsize=10)
    ax.set_xlabel('$x_t$', fontsize=9)
    ax.set_ylabel('$y_t$', fontsize=9)
    ax.tick_params(labelsize=8)

plt.tight_layout(rect=[0, 0, 1, 0.94])

# Save to the same directory as this script for easy replication
out_dir = os.path.dirname(os.path.abspath(__file__)) if '__file__' in dir() else '.'
plt.savefig(os.path.join(out_dir, 'spurious_regression_multiple.png'), dpi=300, bbox_inches='tight')
plt.savefig(os.path.join(out_dir, 'spurious_regression_multiple.pdf'), bbox_inches='tight')
print("\nPlots saved!")

File: Chapter1/Codes_Figures/spurious_regression.R
# =========================================================
# Spurious Regression: Nine Regressions of Independent Random Walks
# =========================================================
# Author:   Alessia Paccagnini
# Textbook: Macroeconometrics
# =========================================================

library(ggplot2)
library(gridExtra)

set.seed(NULL)  # Random seed: we'll try until we find good results

# Output directory: same folder as this script (portable)
out_dir <- tryCatch(dirname(normalizePath(sys.frame(1)$ofile)),
                    error = function(e) getwd())

# --------------------------------------------------------
# Helper: Durbin-Watson statistic
# --------------------------------------------------------
durbin_watson <- function(resid) {
  sum(diff(resid)^2) / sum(resid^2)
}

# --------------------------------------------------------
# Search for 9 pairs with R2 > 0.80 and DW < 0.20
# --------------------------------------------------------
T            <- 300
max_attempts <- 100000
attempt      <- 0
results      <- list()

while (length(results) < 9 && attempt < max_attempts) {
  attempt <- attempt + 1

  x <- cumsum(rnorm(T))
  y <- cumsum(rnorm(T))

  fit   <- lm(y ~ x)
  r2    <- summary(fit)$r.squared
  t_stat <- coef(summary(fit))["x", "t value"]
  dw    <- durbin_watson(residuals(fit))

  if (r2 > 0.80 && dw < 0.20) {
    results[[length(results) + 1]] <- list(
      x      = x,
      y      = y,
      fitted = fitted(fit),
      r2     = r2,
      t      = t_stat,
      dw     = dw
    )
    cat(sprintf("Found pair %d: R2=%.2f, t=%.1f, DW=%.2f\n",
                length(results), r2, t_stat, dw))
  }
}

cat(sprintf("\nTotal attempts: %d\n", attempt))
cat(sprintf("Pairs found: %d\n", length(results)))

# --------------------------------------------------------
# Build one ggplot panel per pair
# --------------------------------------------------------
make_panel <- function(res) {
  df      <- data.frame(x = res$x, y = res$y, fitted = res$fitted)
  df      <- df[order(df$x), ]
  abs_t   <- abs(res$t)
  stars   <- ifelse(abs_t > 3.29, "***",
             ifelse(abs_t > 2.58, "**",
             ifelse(abs_t > 1.96, "*", "")))
  t_sign  <- ifelse(res$t < 0, "-", "")
  ttl     <- sprintf("R2=%.2f, t=%s%.1f%s, DW=%.2f",
                     res$r2, t_sign, abs_t, stars, res$dw)

  ggplot(df, aes(x = x, y = y)) +
    geom_point(alpha = 0.45, size = 1.2,
               colour = "#9B59B6") +
    geom_line(aes(y = fitted), colour = "red", linewidth = 0.9) +
    labs(title = ttl, x = expression(x[t]), y = expression(y[t])) +
    theme_bw(base_size = 9) +
    theme(plot.title = element_text(size = 8, hjust = 0.5))
}

panels <- lapply(results, make_panel)

# --------------------------------------------------------
# Arrange and save
# --------------------------------------------------------
combined <- arrangeGrob(
  grobs = panels,
  ncol  = 3,
  top   = "Nine Regressions of Independent Random Walks\n(All relationships are SPURIOUS)"
)

ggsave(file.path(out_dir, "spurious_regression_multiple.png"),
       combined, width = 14, height = 12, dpi = 300)
ggsave(file.path(out_dir, "spurious_regression_multiple.pdf"),
       combined, width = 14, height = 12)

cat("\nPlots saved!\n")
